skfda/inference/hotelling/hotelling.py

```diff
@@ -0,0 +1,214 @@
+            fd2 (FData): Object containing second sample.
+            weights (numpy.array, optional): Weights matrix. If no value
+                is passed then uses Gram matrix if data is in basis
+                representation. Identity matrix is used for discretized data.
+
+        Returns:
+            The value of the statistic.
+
+        Raises:
+            TypeError.
+
+        Examples:
+
+            >>> from skfda.inference.hotelling import hotelling_t2
+            >>> from skfda.representation import FDataGrid, basis
+
+            >>> fd1 = FDataGrid([[1, 1, 1], [3, 3, 3]])
+            >>> fd2 = FDataGrid([[3, 3, 3], [5, 5, 5]])
+            >>> '%.2f' % hotelling_t2(fd1, fd2)
+            '2.00'
+            >>> fd1 = fd1.to_basis(basis.Fourier(n_basis=3))
+            >>> fd2 = fd2.to_basis(basis.Fourier(n_basis=3))
+            >>> '%.2f' % hotelling_t2(fd1, fd2)
+            '2.00'
+
+        References:
+            [1] A. Pini, A. Stamm and S. Vantini, "Hotelling's t2 in
+            separable hilbert spaces", *Jounal of Multivariate Analysis*,
+            167 (2018), pp.284-305.
+
+        """
+    if not isinstance(fd1, FData):
+        raise TypeError("Argument type must inherit FData.")
+
+    if not isinstance(fd2, type(fd1)):
+        raise TypeError("Both samples must be instances of the same type.")
+
+    n1, n2 = fd1.n_samples, fd2.n_samples  # Size of each sample
+    n = n1 + n2  # Size of full sample
+    m = fd1.mean() - fd2.mean()  # Delta mean
+
+    if isinstance(fd1, FDataBasis):
+        if fd1.basis != fd2.basis:
+            raise ValueError("Both FDataBasis objects must share the same "
+                             "basis.")
+        # When working on basis representation we use the coefficients
+        m = m.coefficients[0]
+        k1 = np.cov(fd1.coefficients, rowvar=False)
+        k2 = np.cov(fd2.coefficients, rowvar=False)
+        # If no weight matrix is passed, then we compute the Gram Matrix
+        if weights is None:
+            weights = fd1.basis.gram_matrix()
+        weights = np.sqrt(np.abs(weights))  # TODO
+    else:
+        # Working with standard discretized data
+        m = m.data_matrix[0, ..., 0]
+        k1 = fd1.cov().data_matrix[0, ..., 0]
+        k2 = fd2.cov().data_matrix[0, ..., 0]
+
+    m = m.reshape((-1, 1))  # Reshaping the mean for a proper matrix product
+    k_pool = ((n1 - 1) * k1 + (n2 - 1) * k2) / (n - 2)  # Combination of covs
+
+    if isinstance(fd1, FDataBasis):
+        # Product of pooled covariance with the weights and Moore-Penrose inv.
+        k_inv = np.linalg.pinv(np.linalg.multi_dot([weights, k_pool, weights]))
+        k_inv = weights.dot(k_inv).dot(weights)
+    else:
+        # If data is discrete no weights are needed
+        k_inv = np.linalg.pinv(k_pool)
+
+    return n1 * n2 / n * m.T.dot(k_inv).dot(m)[0][0]
+
+
+def hotelling_test_ind(fd1, fd2, n_reps=None, random_state=None,
```

Comment:
```suggestion
def hotelling_test_ind(fd1, fd2, *, n_reps=None, random_state=None,
```

Reply:
Yeah, but here we want `n_reps` and `random_state` to be keyword-only args.

Reply:
Changed